Add rainbow effect to led strip and cleanup

ASDosjani opened the pull request:
Betaflight PR: https://github.com/betaflight/betaflight/pull/12323

Demonstration: https://youtu.be/i4qeOpRaKrM
New cli commands:
`ledstrip_rainbow_delta` - delta between the leds in hsv color
`ledstrip_rainbow_freq` - frequency of hsv change

Try to change these values with the `set` command.

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

McGiverGim commented:
Don't modify translations files. Only English. The others will be modified from Crowdin automatically, and the PR is simpler with less modified files.

McGiverGim commented on locales/en/messages.json:
```diff
@@ -3490,17 +3490,17 @@
         "description": "Message in the wiring modes in Led Strip"
     },
-    "ledStripVtxFunction": {
+    "ledStripLarsonOverlay": {
         "message": "Larson scanner"
     },
```
Any change in the messages files will force the translators to traslate them again. I don't use leds, but I suppose the change is because it describes better what is it. Am I right?
I recommend for any string added/modified adding a description that will be seen for translators:
```json
"ledStripLarsonOverlay": {
        "message": "Larson scanner",
        "description": "One of the modes for the LED strip feature"
},
```

ASDosjani replied:
Yeah, it describes the function better, I think they leaved this translation key because legacy support.

McGiverGim commented on locales/en/messages.json:
```diff
@@ -3490,17 +3490,17 @@
+    "ledStripLarsonOverlay": {
         "message": "Larson scanner"
     },
-    "ledStripBlinkTitle": {
-        "message": "Blink"
-    },
```
Why is this deleted? I suppose now is the new `ledStripModifierTitle`?

ASDosjani replied:
It's not used anymore.
Before:
![Screenshot_2023-02-10-13-28-35-140-edit_com betaflight betaflight_configurator_debug](https://user-images.githubusercontent.com/62965528/218092815-88e8d717-bef2-4673-8f68-bb4d01c42c32.jpg)

After:
![Screenshot_2023-02-10-13-27-45-971-edit_com betaflight betaflight_configurator_debug](https://user-images.githubusercontent.com/62965528/218092842-fac0ce45-e99b-4cef-8656-327f3f1dd7c6.jpg)

McGiverGim commented on locales/en/messages.json:
```diff
@@ -3490,17 +3490,17 @@
     },
-    "ledStripBlinkLandingOverlay": {
-        "message": "Blink on landing"
+    "ledStripModifierTitle":{
+        "message": "Modifiers"
+    },
+    "ledStripRainbowOverlay": {
+        "message": "Rainbow"
     },
```
Add the same description tag that commented before, please.

ASDosjani replied:
Wdym?

ASDosjani replied:
> Add the same description tag that commented before, please.

I think it's pretty obvious what this string for, don't know what should I write in the description.

McGiverGim replied:
If you see only ”Rainbow" without any other context you know what is it for? That is what see the translators. They don't know this is for LED feature. For this reason I suggested in the other change something generic like "One of the modes for the LED strip feature". This helps a lot the translators to know where the string will be used.

ASDosjani replied:
Done, I added description in the English file.

McGiverGim commented on src/js/injected_methods.js:
```diff
@@ -49,6 +61,16 @@ DataView.prototype.readU32 = function() {
 
+DataView.prototype.readU64 = function() {
+    if (this.byteLength >= this.offset+8) {
+        const right = this.readU32();  //64bit not working properly in js, have to split into two 32bit
```
Why is this not working? This only happens in Betaflight or there is some info out about this?

ASDosjani replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER

I don't wanted to use BigInt because I would have to duplicate some functions then.

McGiverGim replied:
Ok, it's simply the size of the data, not any kind of problem or bug. I don't know if have a "false" function for that is good, or simply call twice the readU32 directly when reading the MSP command as we do for other similar data (I think remember the flight modes maybe?). I think is more clear. @haslinghuis what do you think?

haslinghuis replied:
Agree with using a 2 x 32 bit struct to complement the 32 and 64 led_strip.

ASDosjani replied:
> Agree with using a 2 x 32 bit struct to complement the ledstrip_32 and ledstrip_64 feature.

So is it good now or should I change it?
I don't understand exactly.

haslinghuis replied:
Please use 32 bit data structure. With ledstrip 64 enabled we can use a second 32 bit structure.
Hardest part is remaining backwards compatibility with previous versions.

ASDosjani replied:
> Please use 32 bit data structure. With ledstrip 64 enabled we can use a second 32 bit structure. Hardest part is remaining backwards compatibility with previous versions.

I think you misunderstand the problem. It's not related to led strip length (32 or 64), instead it's the led config length, each configured led has this which contains it's configuration. Previously it was 32bit long but with the new feature I couldn't fit into this so I had to expand it to 64 bit.

The problem is that JavaScript can't use 64 bit numbers only 53 bit so I call the the readU32 twice and return them in an array. I added readU64 where this is happening. It's for clearity and code readability.
I think @McGiverGim wants me not to implement this readU64 and use twice readU32 directly in the code.

I already fixed the backward compatible problem.

haslinghuis replied:
Agree with @McGiverGim Please use twice U32

ASDosjani replied:
Alright, I change it.
Limon added the 64 led feature to cloud build.

McGiverGim commented on src/js/msp/MSPHelper.js:
```diff
@@ -1169,13 +1169,12 @@ MspHelper.prototype.process_data = function(dataHandler) {
+                let ledCount = (data.byteLength - 2) / 8;
 
                 for (let i = 0; i < ledCount; i++) {
+                    const mask = data.readU64();
```
What happens if LED64 is not enabled? This works or crashes?

ASDosjani replied:
It's not related to led count, both works. The 64 bit is the led configuration size.
This is the reason why I had to change it to 64 bit:
https://github.com/betaflight/betaflight/pull/12323#issue-1572957564

McGiverGim replied:
But will work too with older firmware?

ASDosjani replied:
Probably not, I should bind it to firmware version.

McGiverGim replied:
This is easier to do with two 32 bits fields. One for led_low, and the other for led_high that will be zero for older version and only read for new version.

ASDosjani replied:
I think this 64 bit function can be useful in the future.

McGiverGim reviewed (changes requested):
Some comments, I don't have LED to test it, so I can't confirm that this works/don't breaks anything.

ASDosjani commented:
@McGiverGim I made it backward compatible and replaced the `ledStripLarsonOverlay` in all translation files. I don't know what was `ledStripModifierTitle`, I removed that. I also changed the api version to 1.46.0.
Please review it if you have time.

McGiverGim reviewed (changes requested):
I think you have modified the translations files by error again. They will be modified from Crowdin  when this PR is merged.

McGiverGim commented:
> @McGiverGim I made it backward compatible and replaced the `ledStripLarsonOverlay` in all translation files. I don't know what was `ledStripModifierTitle`, I removed that. I also changed the api version to 1.46.0. Please review it if you have time.

The modifications to translations files will be lost. We don't upload them to Crowdin, and we overwrite them with the contents of Crowdin. For this reason we ask to not modify them to the developers. Only the English file, is the only one that we upload to Crowdin.

ASDosjani commented:
> I think you have modified the translations files by error again. They will be modified from Crowdin when this PR is merged.

Does Crowdin know that I changed the name of a key and not adding an extra that translators should translate again?
I think it's easier to fix this in this PR.

ASDosjani commented:
> > @McGiverGim I made it backward compatible and replaced the `ledStripLarsonOverlay` in all translation files. I don't know what was `ledStripModifierTitle`, I removed that. I also changed the api version to 1.46.0. Please review it if you have time.
> 
> The modifications to translations files will be lost. We don't upload them to Crowdin, and we overwrite them with the contents of Crowdin. For this reason we ask to not modify them to the developers. Only the English file, is the only one that we upload to Crowdin.

Fixed.

haslinghuis commented on src/css/tabs/led_strip.less:
```diff
@@ -150,14 +150,14 @@
 			margin-left: -9px;
 		}
 	}
-	.gPoint.function-n {
-		.overlay-n {
```
Needed for backward compatibility

ASDosjani replied:
I think we don't need this, the configurator only compatible from bf 4.0+ and I just checked and bf 4.0 doesn't contain this overlay.
It's even hidden in the process_html function in `led_strip.js` line 88 and only occurs once in the file, there.
So I think it's not supported now anyway.

haslinghuis commented on src/js/msp/MSPHelper.js:
```diff
@@ -20,7 +20,7 @@ import { OSD } from "../tabs/osd";
 const ledDirectionLetters    = ['n', 'e', 's', 'w', 'u', 'd'];      // in LSB bit order
 const ledBaseFunctionLetters = ['c', 'f', 'a', 'l', 's', 'g', 'r']; // in LSB bit
-let ledOverlayLetters        = ['t', 'o', 'b', 'v', 'i', 'w']; // in LSB bit
+let ledOverlayLetters        = ['t', 'y', 'o', 'b', 'v', 'i', 'w']; // in LSB bit
```
Needed for backward compatibility

ASDosjani replied:
I handle it here:
![image](https://user-images.githubusercontent.com/62965528/221978273-8548b778-528b-459c-a3c8-3479aaf06aac.png)

haslinghuis commented on src/js/msp/MSPHelper.js:
```diff
@@ -2481,25 +2526,49 @@ MspHelper.prototype.sendLedStripConfig = function(onCompleteCallback) {
-        }
 
-        mask |= (led.color << 18);
+            mask |= (led.color << 19);
```
Is this a bug?

haslinghuis commented on src/js/msp/MSPHelper.js:
```diff
@@ -2481,25 +2526,49 @@ MspHelper.prototype.sendLedStripConfig = function(onCompleteCallback) {
+            for (let directionLetterIndex = 0; directionLetterIndex < led.directions.length; directionLetterIndex++) {
+                const bitIndex = ledDirectionLetters.indexOf(led.directions[directionLetterIndex]);
+                if (bitIndex >= 0) {
+                    mask |= bit_set(mask, bitIndex + 23);
```
23 instead of 22, a bug?

ASDosjani replied:
No, these changes are intended. The overlay is expanded from 6bit to 7bit and I had to shift every offset after them. That's why I asked if I should allocate more like 8-9 bits for the overlay and function.
I will update the PR with an illustration and then you will understand how it works.

ASDosjani replied:
@haslinghuis wait a bit, I might be stupid. I have to review my code. I might used wrong numbers.

ASDosjani replied:
@haslinghuis see in the MSPHelper.js file line 1169.

haslinghuis commented on src/js/msp/MSPHelper.js:
```diff
@@ -1164,51 +1164,103 @@ MspHelper.prototype.process_data = function(dataHandler) {
+                        };
+
+                        FC.LED_STRIP.push(led);
+                        }
+                    }
```
Reduce Indentation :)

nerdCopter reviewed (approved):
* i've seen video & reports of it working, i cannot test personally, but based on feedback i'll add my approval.

nerdCopter commented:
* please marked any items as resolved if they are indeed resolved.
